File: LeetCode/Medium/Number Of Closed Island/Java/numberOfClosedIsland.java
class Solution {
    int m, n;
    int[][] dir = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    public int closedIsland(int[][] grid) {
        m = grid.length;
        n = grid[0].length;
        boolean[][] visited = new boolean[m][n];
        int cnt=0;
        for (int i=1; i<m-1; i++) {
            for (int j=1; j<n-1; j++) {
                if (grid[i][j] == 0 && !visited[i][j]) {
                    if (dfs(grid, i, j, visited)){
                        cnt++;
                    }
                }
            }
        }
        return cnt;
    }

    private boolean dfs(int[][] grid, int i, int j, boolean[][] visited) {
        if (i==0 || j==0 || i==m-1 || j==n-1 || visited[i][j]) return false;

        visited[i][j] = true;
        boolean ans = true;
        for (int k=0; k<4; k++) {
            int x = i+dir[k][0];
            int y = j+dir[k][1];
            if (x >=0 && y>=0 && grid[x][y]==0 && !visited[x][y])
                if (!dfs(grid, x, y, visited)) ans = false;
        }
        return ans;
    }
}

File: LeetCode/Medium/Number Of Closed Island/Go/number_of_closed_island.go
package main

import "fmt"

func number_of_closed_island(grid [][]int) int {
	n := len(grid)
	m := len(grid[0])
	visited := make([][]bool, n)
	for i := 0; i < n; i++ {
		visited[i] = make([]bool, m)
	}

	closed_island := 0
	for i := 1; i < n-1; i++ {
		for j := 1; j < m-1; j++ {
			if !visited[i][j] && grid[i][j] == 0 {
				if dfs(grid, &visited, i, j, n, m) {
					closed_island++
				}
			}
		}
	}
	return closed_island
}

var directions = [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

func dfs(grid [][]int, visited *[][]bool, i, j, n, m int) bool {
	if i == 0 || j == 0 || j == m-1 || i == n-1 {
		return false
	}
	(*visited)[i][j] = true
	ans := true
	for k := 0; k < 4; k++ {
		x, y := i+directions[k][0], j+directions[k][1]
		if x >= 0 && y >= 0 && x < n && y < m && grid[x][y] == 0 && !(*visited)[x][y] {
			if !dfs(grid, visited, x, y, n, m) {
				ans = false
			}
		}
	}
	return ans
}

func main() {
	fmt.Println(number_of_closed_island([][]int{
		{1, 1, 1, 1, 1, 1, 1, 0},
		{1, 0, 0, 0, 0, 1, 1, 0},
		{1, 0, 1, 0, 1, 1, 1, 0},
		{1, 0, 0, 0, 0, 1, 0, 1},
		{1, 1, 1, 1, 1, 1, 1, 0},
	}))
}
